### ISSUE TITLE
contrib: makeseeds.py improvements

### ISSUE THREAD

**Issue by laanwj** [bitcoin/bitcoin#17020]
Some suggestions for improvements, or at least for discussion between major releases:

- [x] Read [suspicious hosts](https://github.com/bitcoin/bitcoin/pull/16999#issuecomment-536869449)  from a file (done in #17823, removed completely in #24862)
- [x] [Determinism](https://github.com/bitcoin/bitcoin/pull/16999#issuecomment-537070141) could be improved (this is useful for cross-checking)
    - [x] Read ASNs from a file (Done in #24864)
    - [x] Per-ASN limit [isn't consistently enforced?](https://github.com/bitcoin/bitcoin/pull/16999#issuecomment-537173086)
- [x] [A max of 2 seeds per ASN seems a bit strict](https://github.com/bitcoin/bitcoin/pull/16999#issuecomment-536999544) at lest for IPv6 (done in #24818)
- [x] [Should we bump MIN_BLOCKS?](https://github.com/bitcoin/bitcoin/pull/16999#issuecomment-536999544) (done in #24818)
- [x] Some [typos](https://github.com/bitcoin/bitcoin/pull/16999#pullrequestreview-295794240) in the comments
- [x] Handle and hardcode TorV3 peers (after #19954), see #20239
- [ ] Bring back onion functionality past TorV3 switch
    - [ ] We need a source of non-hardcoded V3 peers, currently the only ones are hardcoded and the seeder hasn't been updated to crawl v3 nodes yet (see sipa/bitcoin-seeder#92)
    - [ ] Consider increasing uptime requirements for Tor nodes from 10% as there [appear to be enough](#issuecomment-718622693)

#7398 was a previous attempt to improve this script, but was abandoned, maybe some changes are useful.

**Comment by MarcoFalke** [comment]
good first issue? :trollface:

**Comment by brakmic** [comment]
Hi, 

Would it be acceptable to get the **suspicious hosts** from a URL instead of a file?
I expanded the makeseeds.py with a small example that gets them from Greg Maxwell's page.

Here's the commit: https://github.com/brakmic/bitcoin/commit/2e6c627cd8b8c69e46914b3a6181eeb2a7e77ccd

Regards,

**Comment by laanwj** [comment]
> Would it be acceptable to get the suspicious hosts from a URL instead of a file?
I expanded the makeseeds.py with a small example that gets them from Greg Maxwell's page.

I don't think the script should fetch from a URL automatically (this interferes with deterministic use), but telling the user to get it themselves is fine.

**Comment by sanjaykdragon** [comment]
Updated your first point:
"Read suspicious hosts (and ASNs) from a file instead of hardcoding"

**Comment by vasild** [comment]
> Would it be acceptable to get the suspicious hosts from a URL instead of a file?

A file that comes with the release is signed. Because the release is made e.g. 1 month ago the user knows that this same exact file has been published 1 month ago and if there was some problem with it then somebody would have complained. Accessing a local file does not depend on a centralized service running (the web server).

A file that has to be downloaded is not signed (yes, we can sign it with the same keys as the release and make bitcoind verify the signature). The user never knows when the file was created, maybe the web server is supplying one content to one user and another content to another user. If the web server just bricks, then the file is not accessible.

**Comment by laanwj** [comment]
Now that (since #20237) we get the full number of Tor seeds (512), we could look into increasing, or at least normalizing the uptime requirement for onions:
```python
    # Require at least 50% 30-day uptime for clearnet, 10% for onion.
    req_uptime = {
        'ipv4': 50,
        'ipv6': 50,
        'onion': 10,
    }
```

**Comment by prayank23** [comment]
A. Filtering hosts with multiple ports can be removed IMO:

https://github.com/bitcoin/bitcoin/blob/c44e734dca64a15fae92255a5d848c04adaad2fa/contrib/seeds/makeseeds.py#L215


B. Tor v3 can also be included in the results.

C. Recent observation which can be confirmed with:

```
wget https://gitlab.com/api/v4/projects/33695681/packages/generic/nrich/0.1.1/nrich_0.1.1_amd64.deb
sudo dpkg -i nrich_0.1.1_amd64.deb
host -t a seed.bitcoin.sipa.be | sed -e 's/seed.bitcoin.sipa.be has address //g' | nrich -
```

Possible reasons for vulnerable machines used for bitcoin nodes:

1. False positives
2. Users not aware or don't care
3. Attackers prefer using these for better results
4. Honeypots
5. Other reasons

Leaving 1 which won't be true for all the results, filtering such nodes in `makeseeds.py` should make sense. Below is an example for one IP copied from [`suspicious_hosts.txt`](https://github.com/bitcoin/bitcoin/blob/master/contrib/seeds/suspicious_hosts.txt)


```python
ip = '88.198.17.7'

url = 'https://internetdb.shodan.io/' + ip
response = requests.get(url)

if response.text.find('CVE') != -1:
    print('vulnerable')
```

**Comment by russeree** [comment]
From "Read suspicious hosts (and ASNs) from a file instead of hardcoding (scripts: Read suspicious hosts from a file instead of hardcoding #17823 does this for suspicious hosts)"

I would like to tackle the **(and ASNs) portion of this task**. Firstly in regards to ASNs nothing seems to be hard coded into the generation process currently other than a parse to an external service. Currently each seed ip-address is parsed though xxx.xxx.xxx.xxx.origin.asn.cymru.com. Does this task imply that a complete list of ASNs should be stored locally and updated periodically? The current cost of a complete compressed list of ASNs and their IP address blocks is ~5.5MB. 

Secondly does this mean there should be a suspicious ASNs file? I don't think this is the correct interpretation.

If all my interpretations are incorrect please let me know.

**Comment by RF5** [comment]
I've tried to address some of the more minor points of improvement in #24818 , namely differentiating max seeds per ASN for ipv4 and ipv6, with some more docs and a `MIN_BLOCKS` bump. Any feedback would be appreciated.

**Comment by laanwj** [comment]
> Secondly does this mean there should be a suspicious ASNs file? I don't think this is the correct interpretation.

No, a list of suspicious ASNs has not been proposed. To be honest I don't think that's very useful, neither is the list of suspicious hosts, it makes no sense to maintain that as part of this project. The more general rules "N hosts per ASN" make sense, though.

**Comment by fernandezpablo85** [comment]
The description should be updated.

> Read ASNs from a file (in progress 

Should read:

> Read ASNs from a file (done in  https://github.com/bitcoin/bitcoin/pull/24864)

**Comment by russeree** [comment]
The per ASN limit IS being enforce properly. I wrote a python script to validate the output of node_main.txt and the limits are being enforced. The previous thread is locked.

Output of script. Each line shows the ASN group, IP Type and Count.
https://gist.github.com/russeree/850d299f386fa19a9d819b2887190a2b

Script source
https://github.com/russeree/bitcoin/blob/e9693097fef78a01c47fb244cbebe8f580e179c4/contrib/seeds/test/asn_limit.py

**Comment by willcl-ark** [comment]
> [ ] Bring back onion functionality past TorV3 switch
>    [ ] We need a source of non-hardcoded V3 peers, currently the only ones are hardcoded and the seeder hasn't been updated to crawl v3 nodes yet (see crawler: Collect Tor v3 and I2P addresses? sipa/bitcoin-seeder#92)

I think these two can be checked off post-https://github.com/bitcoin/bitcoin/pull/30008 ?

**Comment by pratikshmalik05** [comment]
can i work on or pr is already done?
